lib/mongoid/association/nested/one.rb

```diff
@@ -50,12 +50,25 @@ def initialize(association, attributes, options)
           @attributes = attributes.with_indifferent_access
           @association = association
           @options = options
-          @class_name = options[:class_name] ? options[:class_name].constantize : association.klass
+          @class_name = class_from(options[:class_name])
```

Comment:
This variable should be refactored to be called `@klass` instead of `@class_name`, because it's not a name String--its an actual `Class` object.